docs/rst/tutorials/micro/memory_management/memory_management.rst

```diff
@@ -0,0 +1,430 @@
+
+.. contents::
+    :depth: 2
+    :local:
+    :backlinks: none
+
+micro-ROS Client provides a full control over the memory usage during build, configuration and runtime.
+This is one of the most important requisites in order to fit in low resources system and to guarantee a hard real-time behaviour.
+
+Memory management in micro-ROS Client can be configured at multiple levels so the user can use different mechanisms for fitting its requirements within the micro-ROS environment.
+
+In general, micro-ROS by default will use:
+
+* Static allocated memory in **build time**
+* A bounded stack memory consumption
+* A bounded amount of **dynamic memory** during entity creation and destruction (configuration time)
+* Zero **dynamic memory** during runtime
+
+
+.. TODO(pgarrido): Link to profiling article
+
+
+During a micro-ROS Client application development the user is able configure the memory at multiple level.
+Along the following sections, those levels are analyzed in detail.
+
+.. _tutorials_micro_memory_management_allocators:
+
+Allocators
+----------
+
+As in the ROS 2 stack, in the micro-ROS stack the dynamic memory allocators can be customized at runtime.
+By default those allocators relies on the ``libc`` implementation of ``malloc``, ``calloc``, ``realloc`` and ``free`` functions.
+However, in some platforms those functions are not available or not encouraged to be used and they can be replaced by platform specific functions.
+One example of this situation is `FreeRTOS allocators <https://www.freertos.org/a00111.html>`_.
+
+An example on how to set custom allocators at runtime when using ``rcutils`` API is:
+
+.. code-block:: c
+
+  allocator = rcutils_get_zero_initialized_allocator();
+  allocator.allocate = custom_allocate;
+  allocator.deallocate = custom_deallocate;
+  allocator.reallocate = custom_reallocate;
+  allocator.zero_allocate = custom_zero_allocate;
+
+  rcutils_set_default_allocator(&allocator);
+
+A reference implementation of those allocators is:
+
+.. code-block:: c
+
+  void * custom_allocate(size_t size, void * state)
+  {
+    // Allocate and return a memory chunk of `size` bytes.
+  }
+
+  void custom_deallocate(void * pointer, void * state)
+  {
+    // Deallocate memory chunk pointed by `pointer`.
+  }
+
+  void * custom_reallocate(void * pointer, size_t size, void * state)
+  {
+    // Reallocate memory chunk pointed by `pointer` to `size` bytes.
+  }
+
+  void * custom_zero_allocate(size_t number_of_elements, size_t size_of_element, void * state)
+  {
+    // Allocate and return a memory chunk of `number_of_elements * size_of_element` bytes, filled with zeros.
+  }
+
+One example implementation of the most basic allocator that targets platforms where no ``libc`` allocators are available is:
+
+.. note::
+
+  This is a naive implementation of an allocator that does not allows memory deallocation.
+  User shall be aware of using `free_all_heap()` when the micro-ROS entities are no longer required.
```

Comment:
Maybe place this note below the code?

Reply:
Fixed here: 9463bbfeeccd4acfd5fba0677ba74ec79a2782d3